staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filter.go

```diff
@@ -0,0 +1,273 @@
+	"k8s.io/apiserver/pkg/util/flowcontrol/metrics"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+
+	rmtypesv1alpha1 "k8s.io/api/flowcontrol/v1alpha1"
+	rmclientv1alpha1 "k8s.io/client-go/kubernetes/typed/flowcontrol/v1alpha1"
+	rmlistersv1alpha1 "k8s.io/client-go/listers/flowcontrol/v1alpha1"
+)
+
+// Interface defines how the request-management filter interacts with the underlying system.
+type Interface interface {
+	// Wait decides what to do about the request with the given digest
+	// and, if appropriate, enqueues that request and waits for it to be
+	// dequeued before returning.  If `execute == false` then the request
+	// is being rejected.  If `execute == true` then the caller should
+	// handle the request and then call `afterExecute()`.
+	Wait(ctx context.Context, requestDigest RequestDigest) (execute bool, afterExecute func())
+
+	// Run monitors config objects from the main apiservers and causes
+	// any needed changes to local behavior
+	Run(stopCh <-chan struct{}) error
+}
+
+// This request filter implements https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
+
+// requestManagerState is the variable state that this filter is
+// working with at a given point in time.
+type requestManagerState struct {
+	// flowSchemas holds the flow schema objects, sorted by increasing
+	// numerical (decreasing logical) matching precedence.  Every
+	// FlowSchema in this slice is immutable.
+	flowSchemas apihelpers.FlowSchemaSequence
+
+	// priorityLevelStates maps the PriorityLevelConfiguration object
+	// name to the state for that level.  Every field of every
+	// priorityLevelState in here is immutable.  Every name referenced
+	// from a member of `flowSchemas` has an entry here.
+	priorityLevelStates map[string]*priorityLevelState
+}
+
+// priorityLevelState holds the state specific to a priority level.
+type priorityLevelState struct {
+	// config holds the configuration after defaulting logic has been applied.
+	// Exempt may be true while there are queues, in the case of a priority
+	// level that recently switched from being non-exempt to exempt and whose
+	// queues are still draining.
+	// If there are queues then their parameters are here.
+	config rmtypesv1alpha1.PriorityLevelConfigurationSpec
+
+	// qsConfig holds the QueueSetConfig derived from `config` if
+	// config is not exempt, garbage otherwise
+	qsConfig fq.QueueSetConfig
+
+	queues fq.QueueSet
+
+	// Non-nil while waiting for queues to drain.
+	// May be non-nil only if queues is non-nil.
+	// May be non-nil while exempt.
+	emptyHandler *emptyRelay
+}
+
+// requestManager holds all the state and infrastructure of
+// this filter
+type requestManager struct {
+	// wg is kept informed of when goroutines start or stop or begin or end waiting
+	wg counter.GoRoutineCounter
```

Comment:
It doesn't look like this needs to be stored, it's never referenced after construction? `wg` is no longer an accurate name.

Reply:
good catch